feat: alert component

abhinav700 opened the pull request:
## 📝 Description

- closes: #2250
- closes: https://github.com/nextui-org/nextui/discussions/313

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a new Alert component for user notifications and alerts.
	- Added support for customizable alert configurations including colors, radius, and closability.
	- New icons for different alert types (Success, Danger, Info, Warning) implemented.

- **Documentation**
	- New README and comprehensive documentation for the Alert component added.
	- Expanded documentation structure to include the Alert component.

- **Tests**
	- Implemented a test suite to ensure the functionality of the Alert component.
	- Added Storybook stories to showcase various alert configurations.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

wingkwong commented on apps/docs/tsconfig.json:
```diff
@@ -25,6 +25,6 @@
     ".next/types/**/*.ts",
     ".contentlayer/generated"
-  ],
+, "content/docs/components/alert.mdx"  ],
```
what's that for?

abhinav700 replied:
It appeared automatically when I ran pnpm build for working with docs

wingkwong commented on packages/components/alert/src/alert-icons.tsx:
```diff
@@ -0,0 +1,59 @@
+
+export const iconColor = (color: string) => {
+  switch (color) {
+    case "primary":
+      return "#006FEE";
+    case "secondary":
+      return "#9353D3";
+    case "success":
+      return "#17C964";
+    case "warning":
+      return "#F5A524";
+    case "danger":
+      return "#F31260";
+    default:
+      return "white";
```
shouldn't use hardcoded values. Use tw classes instead.

abhinav700 replied:
Ok

wingkwong commented on packages/components/alert/src/alert-icons.tsx:
```diff
@@ -0,0 +1,59 @@
+  }
+};
+
+export const SuccessIcon = () => {
```
for each icon, put it in shared packages I shared

wingkwong commented on packages/components/alert/src/alert.tsx:
```diff
@@ -0,0 +1,86 @@
+
+const closeIconColor = (color: string) => {
+  switch (color) {
+    case "primary":
+      return "#004493";
+    case "secondary":
+      return "#481878";
+    case "success":
+      return "#0E793C";
+    case "warning":
+      return "#936316";
+    case "danger":
+      return "#920B3A";
+    default:
+      return "#71717A";
```
use TW classes

abhinav700 replied:
Ok

wingkwong commented on packages/components/alert/src/use-alert.ts:
```diff
@@ -0,0 +1,103 @@
+
+  ref?: ReactRef<HTMLDivElement | null>;
+
+  title: string;
+  description: ReactNode;
```
add comments

wingkwong commented on packages/components/alert/src/use-alert.ts:
```diff
@@ -0,0 +1,103 @@
+  /**
+   * Ref to the DOM node.
+   */
+
```
remove this line

wingkwong commented on packages/components/alert/src/use-alert.ts:
```diff
@@ -0,0 +1,103 @@
+  title: string;
+  description: ReactNode;
+
+  // whether the alert can be closed by user
```
use the same comment styles

wingkwong commented on packages/components/alert/src/use-alert.ts:
```diff
@@ -0,0 +1,103 @@
+export function useAlert(originalProps: UseAlertProps) {
+  const [props, variantProps] = mapPropsVariants(originalProps, alert.variantKeys);
+
+  //isCloseable is true by default if not provided in props
```
add a space after `//`

wingkwong commented on packages/components/alert/src/alert.tsx:
```diff
@@ -0,0 +1,86 @@
+    title,
+    description,
+    isCloseable,
+    visible,
```
rename to `isVisible`

wingkwong commented on packages/components/alert/src/alert-icons.tsx:
```diff
@@ -0,0 +1,47 @@
+
+export const AlertCloseIcon = (props: {color: string}) => {
+  switch (props.color) {
+    case "primary":
+      return <CloseIcon className={`fill-current text-primary-200`} height={20} width={20} />;
+
+    case "secondary":
+      return <CloseIcon className={`fill-current text-secondary-200`} height={20} width={20} />;
+
+    case "success":
+      return <CloseIcon className={`fill-current text-success-200`} height={20} width={20} />;
+
+    case "warning":
+      return <CloseIcon className={`fill-current text-warning-200`} height={20} width={20} />;
+
+    case "danger":
+      return <CloseIcon className={`fill-current text-danger-200`} height={20} width={20} />;
+
+    default:
+      return <CloseIcon className={`fill-current text-default-400`} height={20} width={20} />;
```
there is no variable used in className. You can simply use string. e.g.

```
className="fill-current text-primary-200"
```

abhinav700 replied:
Must have forgot to change that from my previous code. will make changes.

wingkwong commented on packages/components/alert/src/alert-icons.tsx:
```diff
@@ -0,0 +1,47 @@
+
+export const AlertIcon = (props: {color: string}) => {
+  switch (props.color) {
+    case "primary":
+      return <SuccessIcon className={`fill-current text-primary`} />;
+    case "secondary":
+      return <SuccessIcon className={`fill-current text-secondary`} />;
+    case "success":
+      return <SuccessIcon className={`fill-current text-success`} />;
+    case "warning":
+      return <WarningIcon className={`fill-current text-warning`} />;
+    case "danger":
+      return <DangerIcon className={`fill-current text-danger`} />;
+    default:
+      return <InfoCircleIcon className={`fill-current text-default-foreground`} />;
```
same as above

wingkwong commented on packages/components/alert/src/use-alert.ts:
```diff
@@ -0,0 +1,123 @@
+   */
+  ref?: ReactRef<HTMLDivElement | null>;
+
+  /**
+   * title of the alert message
+   */
+  title: string;
+
+  /**
+   * Main body of the alert message
+   */
+  description: ReactNode;
```
i think `title` should be optional while `description` is mandatory.

abhinav700 replied:
ok. will make the changes.

wingkwong commented on packages/components/alert/src/use-alert.ts:
````diff
@@ -0,0 +1,123 @@
+   * @example
+   * ```ts
+   * <Alert classNames={{
+   *    base:"base-classes", // image classes
````
`// image classes` is wrong

wingkwong commented on packages/components/alert/src/use-alert.ts:
```diff
@@ -0,0 +1,123 @@
+  };
+
+  const {title, description, onClose, isCloseable, ref, classNames} = {
+    /**
+     *  isCloseable is true by default if not provided in props
+     */
+    isCloseable: true,
+
+    /**
+     * By default, onClose simply closes the alert
+     */
+    onClose: handleClose,
```
what's the reason defining there here while they are already in `props`? or you just wanna define the default value? if so, you can simply do in this way

```tsx
  const {title, description, onClose = handleClose, isCloseable = true, ref, classNames} = props;
```

abhinav700 replied:
yeah, I was trying to set defautl values. Will use the syntax you suggested.

wingkwong commented on apps/docs/content/components/alert/usage.ts:
```diff
@@ -0,0 +1,19 @@
+const App = `import {Alert} from "@nextui-org/react";
+
+export default function App() {
+  const title = "Email Sent!!";
+  const description ="You will get a reply soon";
+  return (
+    <div className="flex items-center justify-center w-screen">
+          <Alert title = {title} description = {description}/>
+    </div>
+  );
+}`;
```
please lint the code (check other files as well)

abhinav700 replied:
done. I am using `pnpm lint fix` to do that, is it okay?

wingkwong commented on packages/core/theme/src/components/alert.ts:
```diff
@@ -0,0 +1,94 @@
+        base: "rounded-none",
+      },
+      sm: {
+        base: "rounded-[12px]",
+      },
+      md: {
+        base: "rounded-[16px]",
+      },
+      lg: {
+        base: "rounded-[22px]",
+      },
+      full: {
+        base: "rounded-[9999px]",
```
use TW class. e.g. small, medium, large, full.

wingkwong commented on packages/core/theme/src/components/alert.ts:
```diff
@@ -0,0 +1,94 @@
+
+const alert = tv({
+  slots: {
+    base: ["group flex flex-row w-[342px] h-[68px]  p-[12px]"],
```
1. `group` -> `group/alert` so that it won't conflict with other components.
2. may double check if you really need group here or not

abhinav700 replied:
Yeah we might not need it here. Will check it.

wingkwong commented on packages/core/theme/src/components/alert.ts:
```diff
@@ -0,0 +1,94 @@
+
+const alert = tv({
+  slots: {
+    base: ["group flex flex-row w-[342px] h-[68px]  p-[12px]"],
+    title: ["text-medium font-normal block h-[24px]"],
+    description: ["text-small font-normal h-[20px]"],
+    mainWrapper: ["w-[268px] h-[44px] ml-[20px] flex flex-col box-border items-start"],
+    closeButton: ["w-[24px h-[24px] cursor-pointer relative"],
```
don't use fixed values here.

wingkwong replied:
We don't use the [arbitrary values](https://tailwindcss.com/docs/width#arbitrary-values) but TW classes. e.g. `p-[12px]` -> `p-3`, `h-[20px]` -> `h-5`.

abhinav700 replied:
ok

wingkwong commented on packages/utilities/shared-icons/src/info-circle.tsx:
```diff
@@ -0,0 +1,21 @@
+import React from "react";
+
+import {IconSvgProps} from "./types";
+export const InfoCircleIcon = (
```
add a new line before export (apply to other files)

wingkwong commented on packages/utilities/shared-icons/src/info-circle.tsx:
```diff
@@ -0,0 +1,21 @@
+import React from "react";
```
declare but never read. (please check other files as well)

wingkwong reviewed (changes requested):
please change all `isCloseable ` to `isClosable`

wingkwong reviewed (changes requested):
1. Alert with `isClosable` doesn't close after clicking the close button
2. too much space
![image](https://github.com/user-attachments/assets/c26cf5a3-4fec-4ad2-8df3-a4298c41c9ab)
3. isClosable preview is not rendering in docs, meaning something wrong with your doc code
![image](https://github.com/user-attachments/assets/7abd8b1f-c435-411e-ab02-a3aa2530bf88)
4. please also handle RTL
![image](https://github.com/user-attachments/assets/f0446789-49d2-48ee-8bb0-4959c206384b)
5. in `isClosable`, please put `isClosable = true` and `isClosable = false` next to the Alert component just like other stories to keep it consistent.
6. please check coderabbitai comments as well.

abhinav700 commented:
> 1. Alert with `isClosable` doesn't close after clicking the close button

I think that's true for other components as well, but I will check it.

abhinav700 commented:
^ please also handle RTL

 @wingkwong please can you explain what you mean by this? isn't it correctly displayed

wingkwong commented:
You can see the space between the icon and title is not consistent. 

![image](https://github.com/user-attachments/assets/04f7f61c-5e70-4e62-bcff-55c1f55f6163)

![image](https://github.com/user-attachments/assets/6bb2fa80-7031-4d93-8fbb-288f6ff8e8c3)

abhinav700 commented:
Okay I will fix that.

abhinav700 commented:
is it okay?
![image](https://github.com/user-attachments/assets/30533dcb-34bb-4504-bf67-5ecf8356cc0d)

wingkwong commented:
Yes as long as the space is consistent.

abhinav700 commented:
I have made the changes.
I have ignored some of the changes recommended by coderabbit because they were not consistent with codebase of other components.

Changes made:
1) `isClosable` is false by default (setting it true seemed weird given the prescence of `onClose`)
2) fixed the close button
3) changed the docs for close button
4) added a test case for close button

wingkwong commented on packages/components/alert/src/use-alert.ts:
```diff
@@ -0,0 +1,115 @@
+
+  const handleClose = () => {
+    setIsVisible(false);
+    if (onClose) onClose();
```
```ts
onClose?.();
```

wingkwong commented on packages/components/alert/stories/alert.stories.tsx:
```diff
@@ -0,0 +1,104 @@
+    <div className="flex flex-col">
+      {["default", "primary", "secondary", "success", "warning", "danger"].map((color) => (
+        <div key={color} className="w-full flex justify-between items-center my-3">
+          <h4 className="h4 mx-4 text-md">{color}</h4>
```
apply to all h4 tags

```ts
<span className="mx-4">{color}</span>
```

wingkwong commented on packages/core/theme/src/components/alert.ts:
```diff
@@ -0,0 +1,92 @@
+
+const alert = tv({
+  slots: {
+    base: ["flex flex-row w-[342px] h-17  p-3"],
```
- extra space
- try not to px as unit

wingkwong commented on packages/core/theme/src/components/alert.ts:
```diff
@@ -0,0 +1,92 @@
+    base: ["flex flex-row w-[342px] h-17  p-3"],
+    title: ["text-medium font-normal block h-6"],
+    description: ["text-small font-normal h-5"],
+    mainWrapper: ["w-[268px] h-11 ms-5 flex flex-col box-border items-start"],
```
try not to px as unit

abhinav700 replied:
should I replace as follows:
w-[342px] -> w-114
w-[268px[ -> w-89

abhinav700 commented on packages/components/alert/stories/alert.stories.tsx:
```diff
@@ -0,0 +1,118 @@
+
+const Template = (args) => <Alert {...args} />;
+
+const ColorTemplate = (args) => {
+  return (
+    <div className="flex flex-col">
+      {["default", "primary", "secondary", "success", "warning", "danger"].map((color) => (
+        <div key={color} className="w-full flex justify-between items-center my-3">
+          <h4 className="mx-4 text-md">{color}</h4>
+          <Alert {...args} color={color} />
+        </div>
+      ))}
+    </div>
+  );
+};
+
+const RadiusTemplate = (args) => {
+  return (
+    <div className="flex flex-col">
+      {["none", "sm", "md", "lg", "full"].map((radius) => (
+        <div key={radius} className="w-full flex justify-between items-center my-3">
+          <h4 className="mx-4 text-md">{radius}</h4>
+          <Alert {...args} radius={radius} />
+        </div>
+      ))}
+    </div>
+  );
+};
```
@wingkwong should I implement this change or simply ignore it?

wingkwong replied:
i dont think you need to use h3 or h4. simply span would do. I remember i made the similar comment before

wingkwong commented on packages/components/alert/__tests__/alert.test.tsx:
```diff
@@ -0,0 +1,85 @@
+    expect(closeButton).toBeNull();
+  });
+
+  it("should call the onClose function when clicking on close button", async () => {
```
async not required

wingkwong commented on packages/components/alert/__tests__/alert.test.tsx:
```diff
@@ -0,0 +1,85 @@
+    expect(onClose).toHaveBeenCalled();
+  });
+
+  it("should close the alert when clicking on close button", async () => {
```
async not required

wingkwong reviewed (changes requested):
1. for `apps/docs/content/components/alert/isClosable.ts`, please rename the file to `is-closable.ts` instead to keep the naming consistent.
2. please handle responsive case.

abhinav700 commented:
> 2. please handle responsive case.

By adding different variants for `sm` , `md` and `lg` right?

wingkwong commented:
@abhinav700 you may try by resizing the browser to see and compare with input component.

ryo-manba commented on packages/components/alert/src/alert-icons.tsx:
```diff
@@ -0,0 +1,46 @@
+  WarningIcon,
+} from "@nextui-org/shared-icons";
+
+export const AlertCloseIcon = (props: {color: string}) => {
+  switch (props.color) {
+    case "primary":
+      return <CloseIcon className="fill-current text-primary-200" height={20} width={20} />;
+
+    case "secondary":
+      return <CloseIcon className="fill-current text-secondary-200" height={20} width={20} />;
+
+    case "success":
+      return <CloseIcon className="fill-current text-success-200" height={20} width={20} />;
+
+    case "warning":
+      return <CloseIcon className="fill-current text-warning-200" height={20} width={20} />;
+
+    case "danger":
+      return <CloseIcon className="fill-current text-danger-200" height={20} width={20} />;
+
+    default:
+      return <CloseIcon className="fill-current text-default-400" height={20} width={20} />;
+  }
```
I believe the style for `AlertIcon` should also be defined in `alert.ts`.

abhinav700 replied:
Just to be sure, You only mean the close icon right?

ryo-manba replied:
I think the same approach should be applied to AlertIcon as well.

abhinav700 replied:
okay

abhinav700 replied:
I have implemented these changes.

ryo-manba replied:
Sorry if I wasn't clear. I think `className` should not be defined here.
Instead, the intent is to call it like the following examples.

https://github.com/nextui-org/nextui/blob/765dac53cfe6b04d1b1afda2410afb2ce3844f0f/packages/components/avatar/src/avatar.tsx#L45-L47

https://github.com/nextui-org/nextui/blob/765dac53cfe6b04d1b1afda2410afb2ce3844f0f/packages/components/modal/src/modal-content.tsx#L57-L59

abhinav700 replied:
Just to be sure,
Instead of defining the `className` seperately for `closeIcon` slot, I should define them in `closeButton` itself and avoid the use of `getCloseIconProps`, right?

![image](https://github.com/user-attachments/assets/25cc7c90-d39a-4e41-9478-33e81e43951a)

abhinav700 replied:
I have implemented the changes in my latest commit.
I rmeoved getCloseIconProps. I am calling it in a way similar to avatar component

abhinav700 commented on apps/docs/content/components/alert/custom-impl.ts:
```diff
@@ -0,0 +1,140 @@
+const InfoCircleIcon = `export const InfoCircleIcon = (props) => (
+  <svg
+    fill="none"
+    height="24"
+    viewBox="0 0 24 24"
+    width="24"
+    xmlns="http://www.w3.org/2000/svg"
+    {...props}
+  >
+    <path
+      d="M12 22C17.51 22 22 17.51 22 12C22 6.49 17.51 2 12 2C6.49 2 2 6.49 2 12C2 17.51 6.49 22 12 22ZM12.75 16C12.75 16.41 12.41 16.75 12 16.75C11.59 16.75 11.25 16.41 11.25 16L11.25 11C11.25 10.59 11.59 10.25 12 10.25C12.41 10.25 12.75 10.59 12.75 11L12.75 16ZM11.08 7.62C11.13 7.49 11.2 7.39 11.29 7.29C11.39 7.2 11.5 7.13 11.62 7.08C11.74 7.03 11.87 7 12 7C12.13 7 12.26 7.03 12.38 7.08C12.5 7.1 ...
+    />
+  </svg>
+);`;
```
ignore

abhinav700 commented:
> @abhinav700 you may try by resizing the browser to see and compare with input component.

@abhinav700 I have added breakpoints to change the width and make it more responsive.
can you check it?

wingkwong commented:
@abhinav700 i think the reason is you have the fixed width so it won't shrink when the window is resizing. You may try if you put two alert on the same row and resize the container. Adding breakpoint doesn't solve the full problem since it may not reach the breakpoint.

abhinav700 commented:
> @abhinav700 i think the reason is you have the fixed width so it won't shrink when the window is resizing. You may try if you put two alert on the same row and resize the container. Adding breakpoint doesn't solve the full problem since it may not reach the breakpoint.

I think my original code would work if I replace `w-` with `max-w-`


https://github.com/user-attachments/assets/ba01d876-d782-4dd4-8158-f4cd4e0b3386

wingkwong commented:
Let's discuss in discord.

wingkwong reviewed (changes requested):
for the documentation, please sync with canary branch once and adopt the latest format (see `apps/docs/content/components/accordion`).

abhinav700 commented:
Okay, will look into that

wingkwong commented:
please click re-request review once it's ready

jrgarciadev commented:
Hey @abhinav700 thanks for contributing and creating this component, please check my changes https://github.com/nextui-org/nextui/pull/3982
